### ISSUE TITLE
 SCRIPT5007: Unable to get property 'a' of undefined or null reference 

### ISSUE THREAD

**Issue by Geoffrey-A** [Microsoft/ChakraCore#4266]
Hello,

The error as per title happens with the attached (zipped) HTML file under MS Edge (EdgeHTML versions 16, 15, 14) and IE 11.
It does not produce any error under Firefox or Chrome.

[bug_repro.zip](https://github.com/Microsoft/ChakraCore/files/1480817/bug_repro.zip)

I was not able to test using the pre-built ChakraCore binaries from the release page. I am posting here as indicated on MS Edge issue tracker page for JS language issues.

Best regards,
Geoffrey

**Comment by MikeHolman** [comment]
Looks like JIT bug in flowgraph peeps.

I was able to get repro that doesn't need browser as well:

````
var c;
var blah;
var foo;
(function(f) {
    function b() {
        switch (d().c.b) {
            case 0:
                e(!1);
            case 5E3:
                e(!0)
        }
    }

    function d() {
        return {
            c: {
                b: 0
            },
            face: {
                a: 12.3
            }
        }
    }

    function e(b) {
        var a = d();
        a = null !== a.face && 0 < a.face.a;
        if (!a || b) blah = a
    }
    f.f = b;
    foo = b
})(c || (c = {})); // This is just a sample script. Paste your real code (javascript or HTML) here.
for(var i=0; i<10000; ++i)
foo();
````

**Comment by Geoffrey-A** [comment]
Thank you very much for your work.

I have been able to reproduce the error with your adapted snippet using the latest released Chakra binary. I was also able to confirm that your PR fixes it, by building Chakra with and without your fix. I also checked that, on a debug build, with the added parameters `-maxInterpretCount:1 -maxSimpleJitRunCount:1` the code added to the test suite (`test/Optimizer/fgpeepbug.js`) reproduces the error as well, when your fix is excluded.

I have a few questions regarding the relations between Chakra and the MS browsers:
– The actual code the repro snippet was extracted from did in fact not produce any error on IE11 and EdgeHTML 14. I am guessing this might be because larger codes are less prone to be JITted. Is it correct? Was there a change from EdgeHTML 15 and above that made the JIT be more aggressive?
– Is IE also using the Chakra code from this repo?
– How to identify the version of the Chakra engine being used in a particular release of the Edge/IE browsers?

**Comment by MikeHolman** [comment]
Yes I think you're right that we must be JITing more aggressively in your scenario if issue only started showing up now, because the root problem goes all the way back to IE11. But the heuristics for when/what we JIT are complicated so it's hard for me to know what exactly changed. The only hard and fast rule is the more times you call a function the more likely we are to JIT it. We have made some changes recently with how we JIT loops that contain function calls, as well as function calls inside of try/catch and try/finally statements. Other than that I'm not too sure what changed.

IE is using a much older version of Chakra code, and it isn't getting any more updates except for security fixes. So, for example, this bug will not be fixed in IE.

Unfortunately I don't think the browser says anything about Chakra version number, and I don't know that it is documented anywhere. But I think the versions roughly correspond like this:
EdgeHTML16<->ChakraCore1.8
EdgeHTML15<->ChakraCore1.6
EdgeHTML14<->ChakraCore1.4

**Comment by Geoffrey-A** [comment]
Thank you for the information.
I wondered if Chakra had external parameters (such as maxInterpretCount) that could have been tuned up when it was incorporated into Edge 15.

I was quite surprised to see this bug was hiding for so long in so big browsers.

Seeing it was merged in the release/1.8 branch, I am hopeful to see that bug fix soon in Edge!
